File: src/main/java/com/gitlab/daring/image/config/ConfigDiffBuilder.java
package com.gitlab.daring.image.config;

import com.typesafe.config.Config;
import com.typesafe.config.ConfigValue;
import com.typesafe.config.ConfigValueType;
import java.util.Map;

public class ConfigDiffBuilder {

	public Config build(Config c1, Config c2) {
		Config rc = c1;
		for (Map.Entry<String, ConfigValue> me : c1.root().entrySet()) {
			String p = me.getKey();
			ConfigValue cv = me.getValue();
			if (!c2.hasPath(p)) continue;
			if (cv.equals(c2.getValue(p))) {
				rc = rc.withoutPath(p);
			} else if (cv.valueType() == ConfigValueType.OBJECT) {
				Config sc = build(c1.getConfig(p), c2.getConfig(p));
				rc = rc.withValue(p, sc.root());
			}
		}
		return rc;
	}

}

File: src/main/kotlin/com/gitlab/daring/image/config/ConfigDiffBuilder.kt
package com.gitlab.daring.image.config

import com.typesafe.config.Config
import com.typesafe.config.ConfigValueType

class ConfigDiffBuilder {

    fun build(c1: Config, c2: Config): Config {
        var rc = c1
        for ((p, cv) in c1.root()) {
            if (!c2.hasPath(p)) continue
            if (cv == c2.getValue(p)) {
                rc = rc.withoutPath(p)
            } else if (cv.valueType() == ConfigValueType.OBJECT) {
                val sc = build(c1.getConfig(p), c2.getConfig(p))
                rc = if (sc.isEmpty) rc.withoutPath(p) else rc.withValue(p, sc.root())
            }
        }
        return rc
    }

}